cpp/src/arrow/ipc/options.h

```diff
@@ -67,6 +67,15 @@ struct ARROW_EXPORT IpcWriteOptions {
   /// May only be UNCOMPRESSED, LZ4_FRAME and ZSTD.
   std::shared_ptr<util::Codec> codec;
 
+  /// @brief Always use compression, regardless of the resulting output size
+  ///
+  /// If false, compression will be skipped for body buffers where the expected compressed
+  /// size matches/exceeds the uncompressed size. Otherwise, compression will be applied
+  /// indiscriminately. This option is ignored if no codec was supplied.
+  ///
+  /// This is true by default due to backwards-compatibility concerns.
+  bool compress_always = true;
```

Comment:
We could perhaps get smart and make this an optional floating-point threshold. For example only compress if it would save at least 10% of the buffer size.

On the flip side, compression is typically much slower than decompression, and compression is attempted anyway, so the end-to-end cost might not vary that much. @lidavidm @westonpace Thoughts?

Reply:
A more advanced strategy would be to compress only parts of the buffer and compute a compression ratio on those sampled parts; that's what Dask [does](https://github.com/dask/distributed/blob/0063de53fed5e4e2e409940213c6265867e6635d/distributed/protocol/compression.py#L153).

Reply:
Of course, we don't have to implement an advanced strategy here. However, we must make an API choice and I think it's better to be forward-looking (and implement the heuristic in the most simplistic way, for now).

Reply:
I think the threshold sounds good.

I suppose it may not save encoder time, but it could save some decoder time, right? Else I suppose I don't see the point of the optimization at all.

Reply:
It would save decoder time. But also, if using a sophisticated encoding strategy as mentioned above, it would also save encoder time.